<a href="https://colab.research.google.com/github/AmirJffr/DataScience-and-Machine-Learning-Master-SI/blob/main/Assignements/Part%204/Assignment_part_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### DSML investigation

You are part of the Suisse Impossible Mission Force, or SIMF for short. You need to uncover a rogue agent that is trying to steal sensitive information.

Your mission, should you choose to accept it, is to find that agent before stealing any classified information. Good luck!

# Assignement part four

#### Identifying the suspects' credit score
We received informations that the rogue agent has a *good* credit score.

Our spies at SIMF have managed to collect financial information relating to our suspects as well as a training dataset.

Create a Neural Network over the training dataset `df` to identify which of the suspects have a *standard* `Credit_Mix`.


## Getting to know our data

* `Age`: a user's age

* `Occupation`: a user's employment field

* `Annual_Income`: a user's annual income

* `Monthly_Inh_Salary`: the calculated salary received by a given user on a monthly basis

* `Num_Bank_Accounts`: the number of bank accounts possessed by a given user

* `Num_Credit_Cards`: the number of credit cards a given user possesses

* `Interest_Rate`: The interest rate on those cards (if multiple then it's the average)

* `Num_of_Loans`: The number of loans of each user

* `Delay_from_due_date`: payment tardiness of user

* `Num_of_Delayed_Payment`: the count of delayed payments

* `Changed_Credit_Limit`: NaN

* `Num_Credit_Inquiries`: NaN

* `Credit_Mix`: The user's credit score

* `Outsting_Debt`: Outstanding debt

* `Credit_Utilization_Ratio`: the percentage of borrowed money over borrowing allowance

* `Payment_of_Min_Amount`: does the user usually pay the minimal amount (categorical)

* `Total_EMI_per_month`: Monthly repayments to be made

* `Amount_invested_monthly`: The amount put in an investment fund by the user on a monthly basis

* `Payment_Behaviour`: the user's payment behavior (categorical)

* `Monthly_Balance`: The user's end of the month balance

* `AutoLoan`: If the user has an active loan for their vehicle

* `Credit-BuilderLoan`: If the user has a loan to increase their credit score

* `DebtConsolidationLoan`, `HomeEquityLoan`, `MortgageLoan`, `NotSpecified`, `PaydayLoan`, `PersonalLoan`, `StudentLoan`: different types of loans (categorical features)


In [2]:
# Import required packages

from sklearn.preprocessing import LabelEncoder, OneHotEncoder
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.preprocessing import MinMaxScaler

%matplotlib inline

In [3]:
df = pd.read_csv("https://raw.githubusercontent.com/michalis0/DataScience_and_MachineLearning/master/Assignements/Part%204/data/train_classification.csv", index_col='Unnamed: 0').dropna()
suspects = pd.read_csv("https://raw.githubusercontent.com/michalis0/DataScience_and_MachineLearning/master/Assignements/Part%204/data/suspects.csv", index_col='Unnamed: 0').dropna()
suspects.rename(columns={"Payment_Behaviour": "le_Payment_Behaviour", "Payment_of_Min_Amount": "le_Payment_of_Min_Amount"}, inplace=True)

In [4]:
display(df.head())
print(df.shape)
display(suspects.head())
print(suspects.shape)

Age Occupation  Annual_Income  Monthly_Inh_Salary  Num_Bank_Accounts  \
0   23  Scientist       19114.12         1824.843333                  3   
1   24  Scientist       19114.12         1824.843333                  3   
3   24  Scientist       19114.12         4182.004291                  3   
5   28    Teacher       34847.84         3037.986667                  2   
8   35   Engineer      143162.64         4182.004291                  1   

   Num_Credit_Card  Interest_Rate  Num_of_Loan  Delay_from_due_date  \
0                4              3            4                    3   
1                4              3            4                    3   
3                4              3            4                    4   
5                4              6            1                    3   
8                5              8            3                    8   

   Num_of_Delayed_Payment  ...  Monthly_Balance  AutoLoan Credit-BuilderLoan  \
0                       7  ...       186.266702         1                  1   
1                       9  ...       361.444004         1                  1   
3                       5  ...       343.826873         1                  1   
5                       3  ...       303.355083         0                  1   
8                    1942  ...       854.226027         2                  0   

   DebtConsolidationLoan  HomeEquityLoan MortgageLoan  NotSpecified  \
0                      0               1            0             0   
1                      0               1            0             0   
3                      0               1            0             0   
5                      0               0            0             0   
8                      0               0            0             1   

   PaydayLoan PersonalLoan  StudentLoan  
0           0            1            0  
1           0            1            0  
3           0            1            0  
5           0            0            0  
8           0            0            0  

[5 rows x 29 columns]

(29223, 29)


Age  Annual_Income  Monthly_Inh_Salary  Num_Bank_Accounts  \
0  23.0       19114.12         1824.843333                3.0   
1  24.0       19114.12         1824.843333                3.0   
3  24.0       19114.12         4182.004291                3.0   
5  28.0       34847.84         3037.986667                2.0   
8  35.0      143162.64         4182.004291                1.0   

   Num_Credit_Card  Interest_Rate  Num_of_Loan  Delay_from_due_date  \
0              4.0            3.0          4.0                  3.0   
1              4.0            3.0          4.0                  3.0   
3              4.0            3.0          4.0                  4.0   
5              4.0            6.0          1.0                  3.0   
8              5.0            8.0          3.0                  8.0   

   Num_of_Delayed_Payment  Changed_Credit_Limit  ...  Occupation_Journalist  \
0                     7.0                 11.27  ...                    0.0   
1                     9.0                 13.27  ...                    0.0   
3                     5.0                 11.27  ...                    0.0   
5                     3.0                  5.42  ...                    0.0   
8                  1942.0                  7.10  ...                    0.0   

   Occupation_Lawyer  Occupation_Manager  Occupation_Mechanic  \
0                0.0                 0.0                  0.0   
1                0.0                 0.0                  0.0   
3                0.0                 0.0                  0.0   
5                0.0                 0.0                  0.0   
8                0.0                 0.0                  0.0   

   Occupation_MediaManager  Occupation_Musician  Occupation_Scientist  \
0                      0.0                  0.0                   1.0   
1                      0.0                  0.0                   1.0   
3                      0.0                  0.0                   0.0   
5                      0.0                  0.0                   0.0   
8                      0.0                  0.0                   0.0   

   Occupation_Teacher  Occupation_Writer  userID  
0                 0.0                0.0  317991  
1                 0.0                0.0  241892  
3                 1.0                0.0  303376  
5                 0.0                0.0  761992  
8                 0.0                0.0  373318  

[5 rows x 43 columns]

(714, 43)


In [5]:
df["Credit_Mix"].unique()

array(['Good', 'Standard', 'Bad'], dtype=object)

# 1. Preparing the data
## 1.1 Data cleaning
Consider the dataset loaded into the DataFrame `df` which is *train_classification.csv*. We aim to preprocess this data for model training.

Begin by encoding the categorical variables:
- Apply One-Hot Encoding to `Occupation`
- Apply Label Encoding to `Payment_of_Min_Amount` and `Payment_Behaviour`

*Note: To clearly distinguish your newly encoded columns, especially for label encoding, consider renaming them with a prefix. Please, use `le_Payment_of_Min_Amount` and `le_Payment_Behaviour` for the label-encoded new columns.*

In [6]:
one_hot_encoder = OneHotEncoder(sparse_output=False)
label_encoder = LabelEncoder()

df_encoded = df.copy()

occupation_encoded = one_hot_encoder.fit_transform(df_encoded[['Occupation']])
occupation_df = pd.DataFrame(occupation_encoded, columns=one_hot_encoder.get_feature_names_out(['Occupation']), index=df_encoded.index)
df_encoded = pd.concat([df_encoded, occupation_df], axis=1)

df_encoded['le_Payment_of_Min_Amount'] = label_encoder.fit_transform(df_encoded['Payment_of_Min_Amount'])
df_encoded['le_Payment_Behaviour'] = label_encoder.fit_transform(df_encoded['Payment_Behaviour'])

 After encoding, integrate the newly encoded columns back into a new DataFrame named `df_encoded`, and remove the original columns `Occupation`, `Payment_of_Min_Amount`, and `Payment_Behaviour` to avoid redundancy.

In [7]:
df_encoded.drop(['Occupation', 'Payment_of_Min_Amount', 'Payment_Behaviour'], axis=1, inplace=True)

Finally, display the first few rows of `df_encoded` to verify that the encodings are correctly implemented. This prepared DataFrame will be used for subsequent model training.

In [8]:
df_encoded.head()

Age  Annual_Income  Monthly_Inh_Salary  Num_Bank_Accounts  Num_Credit_Card  \
0   23       19114.12         1824.843333                  3                4   
1   24       19114.12         1824.843333                  3                4   
3   24       19114.12         4182.004291                  3                4   
5   28       34847.84         3037.986667                  2                4   
8   35      143162.64         4182.004291                  1                5   

   Interest_Rate  Num_of_Loan  Delay_from_due_date  Num_of_Delayed_Payment  \
0              3            4                    3                       7   
1              3            4                    3                       9   
3              3            4                    4                       5   
5              6            1                    3                       3   
8              8            3                    8                    1942   

   Changed_Credit_Limit  ...  Occupation_Lawyer Occupation_Manager  \
0                 11.27  ...                0.0                0.0   
1                 13.27  ...                0.0                0.0   
3                 11.27  ...                0.0                0.0   
5                  5.42  ...                0.0                0.0   
8                  7.10  ...                0.0                0.0   

   Occupation_Mechanic  Occupation_MediaManager  Occupation_Musician  \
0                  0.0                      0.0                  0.0   
1                  0.0                      0.0                  0.0   
3                  0.0                      0.0                  0.0   
5                  0.0                      0.0                  0.0   
8                  0.0                      0.0                  0.0   

   Occupation_Scientist  Occupation_Teacher  Occupation_Writer  \
0                   1.0                 0.0                0.0   
1                   1.0                 0.0                0.0   
3                   1.0                 0.0                0.0   
5                   0.0                 1.0                0.0   
8                   0.0                 0.0                0.0   

   le_Payment_of_Min_Amount  le_Payment_Behaviour  
0                         1                     6  
1                         1                     2  
3                         1                     2  
5                         1                     4  
8                         1                     5  

[5 rows x 43 columns]

## 1.2 Dataset splitting and rescaling

To effectively train and validate our model, it is crucial to properly prepare and partition the data. Follow these steps to preprocess and split the dataframe df_encoded into training and test subsets, ensuring that our model can generalize well to new data:

- Set `X` as all columns except `Credit_Mix` and `y` as the dependent feature `Credit_Mix`.
- Apply manually Label Encoding to `y` using the function `.map`and encoding with the following : Good=2, Standard=1, Bad=0.
- Use a `random_state` of 42 to split `X` and the encoded `y` into training (80%) and test sets (20%).
- Normalize `X` using `MinMaxScaler()`.

In [9]:
X = df_encoded.drop('Credit_Mix', axis=1)
y = df_encoded['Credit_Mix'].map({'Good': 2, 'Standard': 1, 'Bad': 0})
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

### 1.2.2 Final touches
Convert the features to torch tensors of type `torch.float` and the labels (dependent variables) to torch tensors of type `torch.long`.

In [10]:
X_train = torch.tensor(X_train, dtype=torch.float)
X_test = torch.tensor(X_test, dtype=torch.float)
y_train = torch.tensor(y_train.values, dtype=torch.long)
y_test = torch.tensor(y_test.values, dtype=torch.long)

# 2 Model preparation:

## 2.1 Define a Neural network model and instantiate it.
Define your neural network model as a class in PyTorch, extending from nn.Module. In the `__init__` method, initialize a linear layer using `nn.Linear()` with specified input and output sizes, and set up `nn.ReLU()` for activation. Implement the forward method to describe how data passes through this layer during the network's forward computation.

Set the following parameters:
* `hidden layer` : 1
* `activation function` : ReLU

In [11]:
class Net(nn.Module):
    def __init__(self, D_in, H, D_out):

        super(Net, self).__init__()
        self.linear1 = nn.Linear(D_in, H)
        self.relu = nn.ReLU()
        self.linear2 = nn.Linear(H, D_out)

    def forward(self, x):

        h_relu = self.relu(self.linear1(x))
        y_pred = self.linear2(h_relu)
        return y_pred

Set `D_in` to the number of features in `X_train` and `D_out` to the number of target variables in `y_train`, then print these dimensions to verify their values.

In [12]:
D_in = X_train.shape[1]
D_out = len(y_train.unique())
print(f"D_in: {D_in}, D_out: {D_out}")

D_in: 42, D_out: 3


\
Initialize the `Net` model with the specified input size `D_in`, 150 hidden units, and output size `D_out`.

In [13]:
model = Net(D_in, 150, D_out)
print(model)

Net(
  (linear1): Linear(in_features=42, out_features=150, bias=True)
  (relu): ReLU()
  (linear2): Linear(in_features=150, out_features=3, bias=True)
)


Let's calculate now how many parameters we have in the model.

In [14]:
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)

num_parameters = count_parameters(model)
print(f"The model has {num_parameters} parameters.")

The model has 6903 parameters.


**Q1. How many parameters does your model have ?**

*Note: Enter an integer (e.g. 355)*

6903 parameters

## 2.2 Finding the best model

Determine the optimal hyper-parameters for your model from the options listed below:

* `criterion` : [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)
* `optimizer` : [Stochastic Gradient Descent (SGD)](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)
* `Epochs`: Test with **150**, **250**, **500**, and **1000** epochs.
* `Learning Rate`: Experiment with learning rates of **0.00005**, **0.001**, **1**, and **10**.

**Evaluation**: Assess your model's performance by measuring its accuracy on the test set.

*Note: Run the code `torch.manual_seed(42)` to ensure consistency across all experiments.*

In [15]:
torch.manual_seed(42)   # Set the seed for reproducibility


### 2.2.1 Automatically Tuning Hyperparameters

In this section, you will automate the process of testing different hyperparameter combinations using a loop, as demonstrated in the lab.

Begin by defining the ranges for the hyperparameters you want to explore:
- **Epochs**: Test with **150**, **250**, **500**, and **1000** epochs.
- **Learning Rate**: Experiment with learning rates of **0.00005**, **0.001**, **1**, and **10**.


In [16]:
epochs_list = [150, 250, 500, 1000]
learning_rates = [0.00005, 0.001, 1, 10]

Next, create a loop that iterates over the list of learning rates, with an inner loop iterating over the list of number of epochs. Inside the inner loop, initialize (define) the model, and also define the optimizer and the loss function (criterion). Then train the model as demonstrated in the lab, and after the training evaluate its performance to obtain the test accuracy for each model. Ensure that you also display the test loss, as you will need it to answer the upcoming questions.

Here’s a reminder of the criterion and optimizer you should use:
- **Criterion**: [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)
- **Optimizer**: [Stochastic Gradient Descent (SGD)](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)

**Hint**: Make sure to define the criterion and optimizer at every iteration (inside the inner loop), otherwise the model might retain previous training states and produce biased results.

(Optional) You can enhance the code by storing the best model along with its best_accuracy and best_params. This way, you'll have an automatic evaluation at the end to identify the best-performing model.



In [17]:
best_accuracy = 0
best_params = {}


for lr in learning_rates:
    for epochs in epochs_list:
        print(f"Training with Learning Rate: {lr}, Epochs: {epochs}")

        model = Net(D_in, 150, D_out)
        criterion = nn.CrossEntropyLoss()
        optimizer = optim.SGD(model.parameters(), lr=lr)

        for epoch in range(epochs):
            optimizer.zero_grad()
            outputs = model(X_train)
            loss = criterion(outputs, y_train)
            loss.backward()
            optimizer.step()

        with torch.no_grad():
            outputs = model(X_test)
            _, predicted = torch.max(outputs.data, 1)
            correct = (predicted == y_test).sum().item()
            accuracy = correct / y_test.size(0)

        print(f"Test Loss: {loss.item():.4f}, Test Accuracy: {accuracy:.4f}")

        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_params = {'learning_rate': lr, 'epochs': epochs}

print(f"\nBest Accuracy: {best_accuracy:.4f}")
print(f"Best Parameters: {best_params}")

Training with Learning Rate: 5e-05, Epochs: 150
Test Loss: 1.0891, Test Accuracy: 0.3097
Training with Learning Rate: 5e-05, Epochs: 250
Test Loss: 1.1044, Test Accuracy: 0.2866
Training with Learning Rate: 5e-05, Epochs: 500
Test Loss: 1.0967, Test Accuracy: 0.3334
Training with Learning Rate: 5e-05, Epochs: 1000
Test Loss: 1.0871, Test Accuracy: 0.4380
Training with Learning Rate: 0.001, Epochs: 150
Test Loss: 1.0919, Test Accuracy: 0.3932
Training with Learning Rate: 0.001, Epochs: 250
Test Loss: 1.0625, Test Accuracy: 0.4539
Training with Learning Rate: 0.001, Epochs: 500
Test Loss: 1.0633, Test Accuracy: 0.4546
Training with Learning Rate: 0.001, Epochs: 1000
Test Loss: 1.0460, Test Accuracy: 0.4544
Training with Learning Rate: 1, Epochs: 150
Test Loss: 0.3816, Test Accuracy: 0.8173
Training with Learning Rate: 1, Epochs: 250
Test Loss: 0.3676, Test Accuracy: 0.8303
Training with Learning Rate: 1, Epochs: 500
Test Loss: 0.3446, Test Accuracy: 0.8366
Training with Learning Rate: 1,

### 2.2.3 Questions

**Q2. What is the test accuracy when we train the model with a learning rate of 0.001 and for 150 epochs? Round your answer to 2 decimal points, e.g., 0.25.**




0.39

**Q3. When using 1000 epochs, which learning rate results in the highest test accuracy?**

*Note: Select among the following answers*

1


**Q4. Is BCELoss a suitable alternative to CrossEntropyLoss for our dataset?**

*Hint: Consider the unique values in the Credit_Mix output variable when answering.*

no, it is better for binary classification problems

### 3. Predict on the Suspects Dataset

Now it's time to use your trained model to make predictions on the suspects dataset!

Please retrain on the full dataset.

Use the following parameters for the model:
- **Hidden layer**: 1 hidden layer with 150 neurons
- **Output layer**: 3 neurons for classification
- **Optimizer**: [Stochastic Gradient Descent (SGD)](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)
- **Criterion**: [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)
- **Iterations**: 1000
- **Learning rate**: 1.0

Ensure consistency by setting the manual seed with `torch.manual_seed(42)` before training.

In [18]:
torch.manual_seed(42)

model = Net(D_in, 150, D_out)
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=1.0)

epochs = 1000


Now, train your model on the training dataset just as you did in section 2.2.1, but you don't have to loop over all the hyper parameters. Ensure that you use the correct number of epochs specified earlier.

In [19]:
for epoch in range(epochs):
    optimizer.zero_grad()
    outputs = model(X_train)
    loss = criterion(outputs, y_train)
    loss.backward()
    optimizer.step()


Before making predictions, confirm that the feature column names in the `suspects` dataset match those expected by the model, particularly ensuring the X features correspond accurately.

In [20]:
suspects_processed = suspects.copy()
suspects_processed = suspects_processed[X.columns]

if list(suspects_processed.columns) == list(X.columns):
    print("Feature columns in suspects dataset match the training data after processing.")
else:
    print("Feature columns do NOT match the training data after processing.")

Feature columns in suspects dataset match the training data after processing.


Then scale your dataset and convert it into a torch tensor of dtype float, similar to the preprocessing done for the training set in section 1.2.

In [21]:
suspects_scaled = scaler.transform(suspects_processed)
suspects_tensor = torch.tensor(suspects_scaled, dtype=torch.float)


Make predictions using the trained model and assign the predicted credit score to each user. Ensure to do the following encoding is used for the predicted categories:
* `0` corresponds to bad credit score,
* `1` corresponds to standard credit score,
* `2` corresponds to good credit score.

Use the predictions to add a new column `credit_score` in the `suspects` dataframe that maps the predicted numerical values to the respective credit score categories.

In [22]:
with torch.no_grad():
    outputs = model(suspects_tensor)
    _, predicted_labels = torch.max(outputs.data, 1)

credit_score_map = {0: 'Bad', 1: 'Standard', 2: 'Good'}
suspects['credit_score'] = predicted_labels.numpy()
suspects['credit_score'] = suspects['credit_score'].map(credit_score_map)




As mentioned earlier, we believe the suspect had a "good" credit score. Review the predictions made by the model, then display how many suspects were categorized under each credit score, and extract the `userID`s of those with a "standard" credit score.

In [23]:
print("Distribution of predicted credit scores among suspects:")
display(suspects['credit_score'].value_counts())

standard_credit_suspects = suspects[suspects['credit_score'] == 'Standard']
print("\nUserIDs of suspects with a 'Standard' credit score:")
display(standard_credit_suspects['userID'].tolist())

Distribution of predicted credit scores among suspects:


credit_score
Good        263
Standard    257
Bad         194
Name: count, dtype: int64


UserIDs of suspects with a 'Standard' credit score:


[660506,
 244162,
 353527,
 531937,
 413250,
 512604,
 334769,
 564061,
 306495,
 647912,
 946059,
 917315,
 504629,
 520229,
 861915,
 308845,
 888670,
 449514,
 867078,
 618568,
 512128,
 160212,
 466624,
 903438,
 777470,
 397912,
 123562,
 939524,
 863210,
 717780,
 863592,
 898254,
 424603,
 599462,
 601265,
 402663,
 551506,
 365406,
 688784,
 829946,
 380835,
 823355,
 456546,
 790046,
 932281,
 571364,
 796491,
 133154,
 430864,
 899581,
 240613,
 126115,
 839591,
 453712,
 259083,
 512432,
 874150,
 484182,
 744124,
 641947,
 781819,
 666304,
 619595,
 103928,
 245167,
 893389,
 953342,
 570734,
 743377,
 510420,
 204180,
 564884,
 937535,
 376743,
 767527,
 225178,
 519735,
 706286,
 934741,
 792996,
 140991,
 641389,
 460980,
 241404,
 325946,
 154013,
 537543,
 418245,
 843485,
 99489,
 231948,
 645264,
 649131,
 509130,
 719655,
 316652,
 726678,
 804662,
 456436,
 378721,
 573810,
 769196,
 804201,
 287607,
 866984,
 791891,
 631142,
 474811,
 409012,
 795804,
 431634,
 8

**Q5.Which of the following suspects have a "good" credit mix according to your model's predictions?**


In [24]:
suspect_ids_to_check = [167822, 761992, 519735, 862880, 409055]
good_credit_suspects = suspects[suspects['credit_score'] == 'Good']
good_suspects_found = good_credit_suspects[good_credit_suspects['userID'].isin(suspect_ids_to_check)]

print("Suspects with a 'Good' credit mix among the options:")
display(good_suspects_found[['userID', 'credit_score']])

Suspects with a 'Good' credit mix among the options:


userID credit_score
5     761992         Good
123   167822         Good
1177  409055         Good

## Your investigation is progressing effectively, and the list of suspects is narrowing down.

**Don't forget to answer the quiz and submit your code on Moodle before the end of the deadline.**